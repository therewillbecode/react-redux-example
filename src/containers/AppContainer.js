import React, { Component } from "react";
import { Route, Router } from "react-router-dom";
import createHistory from "history/createBrowserHistory";
import { connect } from "react-redux";
import auth0 from "auth0-js";

import AssetsContainer from "./AssetsContainer"; //
import AssetItemContainer from "./AssetItemContainer"; //
import App from "../components/App";
import AppHeader from "../components/AppHeader";
import Loading from "../components/Loading";
import RequireAuth from "./HOC/RequireAuth";
import { authSuccess, loggedOut } from "../actions/index";
import { isAuthenticated } from "../selectors/index";

const auth0Domain = "therewillbecode.auth0.com";
const auth0ClientId = "2yqWzBSv0zfQEstOyMveQBu4Rw3bqbiT";
const auth0CallbackUrl = "http://localhost:3000/loading";

export class AppContainer extends Component {
  constructor(props) {
    super(props);
    this.login = this.login.bind(this);
    this.logout = this.logout.bind(this);
    this.handleAuthentication = this.handleAuthentication.bind(this);
    this.isAuthenticated = this.isAuthenticated.bind(this);
    this.history = createHistory();
    this.history.listen(({ pathname, hash }) => {
      if (pathname === "/loading" && /id_token|error/.test(hash)) {
        this.handleAuthentication();
      }
    });

    this.auth0 = new auth0.WebAuth({
      domain: auth0Domain,
      clientID: auth0ClientId,
      redirectUri: auth0CallbackUrl,
      audience: `https://${auth0Domain}/userinfo`,
      responseType: "id_token",
      scope: "openid profile"
    });
  }

  componentDidMount() {
    const idToken = localStorage.getItem("id_token");

    // If we have a token, consider the user to be signed in and update state
    if (idToken) {
      this.props.authSuccess(idToken);
    }
  }

  login() {
    this.auth0.authorize();
  }

  handleAuthentication() {
    //  console.log("called");
    this.auth0.parseHash((err, authResult) => {
      if (authResult && authResult.idToken) {
        this.setSession(authResult);
        this.history.replace("/assets");
      } else if (err) {
        this.history.replace("/");
      }
    });
  }

  setSession({ expiresIn, idToken }) {
    // Set the time that the access token will expire at
    let expiresAt = JSON.stringify(expiresIn * 1000 + new Date().getTime());
    localStorage.setItem("id_token", idToken); // JSON web token - decode on server side
    localStorage.setItem("expires_at", expiresAt);
    this.props.authSuccess(idToken);
    // navigate to the assets route
    this.history.replace("/assets");
  }

  logout() {
    // Clear access token and ID token from local storage
    localStorage.removeItem("access_token");
    localStorage.removeItem("id_token");
    localStorage.removeItem("expires_at");
    this.props.loggedOut();
    // navigate to the home route
    this.history.replace("/");
  }

  isAuthenticated() {
    // Check whether the current time is past the
    // access token's expiry time
    let expiresAt = JSON.parse(localStorage.getItem("expires_at"));
    return new Date().getTime() < expiresAt;
  }

  render() {
    const { isAuthenticated, downloadInventoryCSV } = this.props;

    return (
      <Router history={this.history}>
        <div style={{ width: "100%", height: "100%" }}>
          <AppHeader
            login={this.login}
            logout={this.logout}
            isAuthenticated={isAuthenticated}
            downloadInventoryCSV={downloadInventoryCSV}
          />
          <Route
            exact
            path="/"
            component={() => (
              <App login={this.login} isAuthenticated={isAuthenticated} />
            )}
          />
          <Route path="/assets" component={RequireAuth(AssetsContainer)} />
          <Route
            path="/asset/:id"
            component={RequireAuth(AssetItemContainer)}
          />
          <Route
            path="/loading"
            render={props => {
              return (
                <Loading
                  handleAuthentication={this.handleAuthentication}
                  {...props}
                />
              );
            }}
          />
        </div>
      </Router>
    );
  }
}

const mapStateToProps = state => {
  return {
    isAuthenticated: isAuthenticated(state)
  };
};

const mapDispatchToProps = dispatch => {
  return {
    authSuccess: idToken => {
      dispatch(authSuccess(idToken));
    },
    loggedOut: () => {
      dispatch(loggedOut());
    }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(AppContainer);
