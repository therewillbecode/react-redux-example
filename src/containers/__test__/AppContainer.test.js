/* eslint-disable */
import React from "react";
import ReactShallowRenderer from "react-test-renderer/shallow";

import { AppContainer } from "../AppContainer";

describe("AssetsContainer", () => {
  it("renders correctly", () => {
    const renderer = new ReactShallowRenderer();
    renderer.render(<AppContainer />);
    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });
});
