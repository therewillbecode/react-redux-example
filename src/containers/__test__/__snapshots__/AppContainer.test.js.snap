// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetsContainer renders correctly 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <div
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <withRouter(AppHeader)
      downloadInventoryCSV={undefined}
      isAuthenticated={undefined}
      login={[Function]}
      logout={[Function]}
    />
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      component={[Function]}
      path="/assets"
    />
    <Route
      component={[Function]}
      path="/asset/:id"
    />
    <Route
      path="/loading"
      render={[Function]}
    />
  </div>
</Router>
`;
